web/static/js/components/thesis-editor.js

```diff
@@ -53,50 +46,39 @@ class ThesisEditor extends React.Component {
     this.addPagePressed = this.addPagePressed.bind(this)
     this.deletePagePressed = this.deletePagePressed.bind(this)
     this.pageSettingsPressed = this.pageSettingsPressed.bind(this)
-  }
 
-  warnURLRedirect (url) {
-    if (!url) return
-    if (window.confirm(`This page is set to redirect to ${url}. Follow redirect?`)) {
-      window.location = url
-    }
-  }
-
-  ospryPublicKey () {
-    return thesisContainer.getAttribute('data-ospry-public-key')
-  }
-
-  pathname () {
-    return window.location.pathname
-  }
-
-  pageTitle () {
-    return document.title
-  }
-
-  pageDescription () {
-    const desc = this.pageDescriptionMetaTag()
-    return desc ? desc.content : null
-  }
+    // External editors
+    
+    this.htmlEditor = new HtmlEditor({
+      onChange: () => this.setState({pageModified: true})
+    })
 
-  pageRedirectURL () {
-    return thesisContainer.getAttribute('data-redirect-url')
-  }
+    this.rawHtmlEditor = new RawHtmlEditor({
+      openTray: this.openTray('raw-html'),
+      closeTray: this.trayCanceled
+    })
 
-  pageTemplate () {
-    return thesisContainer.getAttribute('data-template')
-  }
+    this.imageEditor = new ImageEditor({
+      ospryPublicKey: this.external.ospryPublicKey,
+      openTray: this.openTray('image-url'),
+      closeTray: this.trayCanceled
+    })
 
-  pageTemplates () {
-    return thesisContainer.getAttribute('data-templates').split(',').filter((s) => s !== '')
-  }
+    this.textEditor = new TextEditor({
+      onChange: () => this.setState({pageModified: true})
+    })
 
-  dynamicPage () {
-    return thesisContainer.getAttribute('data-dynamic-page')
   }
 
-  pageDescriptionMetaTag () {
-    return document.querySelectorAll('meta[name=description]')[0]
+  openTray (trayType) {
```

Comment:
This function is manually curried to allow for modification of `trayType` within the resulting function.